Rakefile

```diff
@@ -33,3 +33,17 @@ task :cleanup_old_stories, [:number_of_days] => :environment do |_t, args|
   args.with_defaults(number_of_days: 30)
   RemoveOldStories.call(args[:number_of_days].to_i)
 end
+
+desc "Generate Postgres env file"
+namespace :env do
+  task :generate_postgres do
+    File.write("./.postgres.env", `erb .docker/.postgres.env.erb`)
+  end
+end
+
+desc "Generate stringer env file"
+namespace :env do
+  task :generate_stringer: :generate_postgres do
```

Comment:
I'm not sure this syntax is correct, but I wanted to make `env:generate_stringer` depend on `env:generate_postgres`. I tried to follow the syntax in the second line [here](https://ruby.github.io/rake/doc/rakefile_rdoc.html#label-Multiple+Definitions).

For the Docker user, how does one run these rake commands? Which package do we need to install to have `rake` available?

Reply:
(if `env:generate_stringer` depends on `env:generate_postgres`, does it automatically run the dependency? i.e. could we just run `rake env:generate_stringer`? What happens if the `.postgres.env` or `.stringer.env` files already exist, does erb recognize that?)

Reply:
> For the Docker user, how does one run these rake commands? Which package do we need to install to have `rake` available?

This depends on the Ruby gems being installed. Then we can run `bundle exec rails env:generate_stringer`.

> if `env:generate_stringer` depends on `env:generate_postgres`, does it automatically run the dependency?

Yep, it'll execute the dependency first.

> What happens if the `.postgres.env` or `.stringer.env` files already exist, does erb recognize that?

Based on my testing locally, it looks like it will overwrite the file. Should we add a check for `File.exist?` and break?

Reply:
Hmm, depending on Ruby being installed isn't super satisfactory - the whole point of docker is not to need anything spcial installed... let me think about this one.

Yeah, we should not overwrite the file. I think if we use `file` instead of `task` it has that behavior, but I can try with `File.exist` as well.

Reply:
> the whole point of docker is not to need anything special installed.

In this case, Ruby and the required gems should be packed in with the image, so I think it should be okay. These commands will be run at runtime inside the image, right?


> I think if we use `file` instead of `task` it has that behavior

Wow, I didn't know about the `file` tasks. What if we do something like:

```rb
file ".postgres.env" do
  ...
end

file ".stringer.env" do
  ...
end

task generate: [".postgres.env", ".stringer.env"]
```

Reply:
> These commands will be run at runtime inside the image, right?

No, we'd need them to run outside of the image (otherwise there's high potential for data loss).

Reply:
Hmm, well docker doesn't really have much in the way of dependencies, does it? It doesn't seem like `make` is necessarily guaranteed either, is it? Does that leave us with just shell scripting?

Reply:
Alternatively, can we run the command inside the docker image and sync it outside?

Reply:
> It doesn't seem like make is necessarily guaranteed either, is it? Does that leave us with just shell scripting?

Well, it's near-guaranteed. It's not Docker itself but rather the system someone is running docker on that would need `make` or `rake` or whatever command we decide on.

> can we run the command inside the docker image and sync it outside?

I don't think so... we would have to run the docker container with a given command, copy the file outside the container, then re-launch the container. And the first time we run the docker container it would have the "wrong" environment variables (because they aren't defined). This could be done but is less-than-ideal because generally dependencies aren't defined within a container but rather the container definition should be standalone.

Actually, as I was writing the above, maybe we _could_ set it up within the container, and just have the first command in the docker setup be something like `docker run --entrypoint=/bin/sh stringer -- rake:generate_env >> .env`, with `rake:generate` just printing output to stdout. So the instructions would be

```sh
docker run ... >> .env
docker compose up -d stringer
```

I think that'd be a satisfactory conclusion, what do you think? We'd keep the `Rakefile` commands and wouldn't need any external dependencies, not even `make`, on the bare metal system.

Reply:
Could we maybe use a volume for it? My docker understanding is very limited, but I'm imagining adding it to the `docker-compose.yml` file, `.env:.env`. Then maybe we could add the command to generate the file `start.sh`.

One additional consideration is what happens on the off chance somebody is trying to use the docker image on a windows machine. It seems less likely they'll have the build tools we need.

Reply:
I think that might work, but I believe we'd have to ask people to create an empty .env file otherwise docker compose complains. And if the file exists on the host then it won't be overwritten by a docker container, so I don't think the container would be able to write to the host disk.

I still think the best conclusion would be to have a separate entrypoint or separate command that we can run with the container and then export to the .env on the host disk - we wouldn't need anything other than docker on the host.

Reply:
Notes from our face-to-face discussion: https://docs.google.com/document/d/1UP4N0szZdYyjXtA2RucE2MxqEm6ln8dhtHQTaXC-o7g/edit#heading=h.dedb9t7qkmm9